### PR TITLE
Refactor klona usage

### PR DESCRIPTION
**Pull request by mikejolley**
This is a followup to https://github.com/WordPress/gutenberg/pull/43631/files to replace `klona` usage in the test suite, and in the `updateState` function. 

@eason9487 pointed out that the new `klona` replacement I did (replacing clone and setWith) was not compatible. The new code is—see codesandbox here for demo:

https://codesandbox.io/s/quiet-tree-0glhh7?file=/src/index.js

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

Automated tests cover the updateState method. Smoke test only.

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

Removed another dependency 🥳

### CODE REVIEW

File: assets/js/data/utils/update-state.ts [hunk 1]
```diff
@@ -1,33 +1,37 @@
 /**
- * External dependencies
+ * Utility for updating nested state in the path that changed.
  */
-import { klona } from 'klona/json';
+function updateNested< Type >( // The state being updated
```
Comment (opr) [inline assets/js/data/utils/update-state.ts]: I usually saw this done using just `T` as the generic name - @gigitux what do you think of this?

File: assets/js/data/utils/update-state.ts [hunk 2]
```diff
@@ -1,33 +1,37 @@
...
+	state: Type,
+	// The path being updated
+	path: string[],
+	// The value to update for the path
+	value: unknown,
+	// The current index in the path
+	index: number
```
Comment (opr) [inline assets/js/data/utils/update-state.ts]: It would be cool if this defaulted to `0`, that way the consumer doesn't have to specify that.

**Review by opr — APPROVED** [review]
Smoke tests seem fine, I focused on product grid related testing and it was OK. I also tested the full flow, all fine.

File: assets/js/data/utils/update-state.ts (hunk 1 above)
Comment (gigitux) [reply assets/js/data/utils/update-state.ts]: Yes, `T` is the standard for the generic type

Comment (mikejolley) [reply assets/js/data/utils/update-state.ts]: Ok I'll change it. I struggle to read and write generics when they use shorthand instead of `Type` 🙈

**Pull request merged**